Amend MPRIS interface

uhle opened the pull request:
I have fixed and re-enabled `CanPlay`/`CanPause` as well as I have fixed `CanGoNext` (which is now false if it is already the last item playing) and `CanGoPrevious` (which is now false if it is the first item playing). In addition, both properties `CanGoNext` and `CanGoPrevious` are always updated when the play queue changes or the position of the currently selected item (which includes switching to the next song and similar things).

CDrummond commented on dbus/mpris.h:
```diff
@@ -93,11 +93,11 @@ class Mpris : public QObject
     bool CanControl() const { return true; }
-    bool CanPlay() const { return true; }
-    bool CanPause() const { return true; }
+    bool CanPlay() const { return MPDState_Playing!=MPDStatus::self()->state(); }
```
Not sure about this. I cant remember why, but 4 years a go a change was made so that CanPlay/CanPause always returned true - https://github.com/CDrummond/cantata/commit/0e5c172671a59c41466264d812646bd51c515cd8

uhle replied:
That is interesting. Good point. I wasn't aware of this and I should have read the [MPRIS interface specification](https://specifications.freedesktop.org/mpris-spec/latest/Player_Interface.html) more thoroughly. You are right, `CanPlay`/`CanPause` shall be set to true most of the time though. Yet they shall be set to false if there is no current item selected. This is only the case if MPD returns `"stopped"` as its state.
It is also a good idea that you have checked the play queue size for `CanPlay`. This way it is possible to call `Play` via MPRIS even if MPD's state is `"stopped"` (assuming the first item of the play queue is selected). I have not tested my previous changes with an empty play queue where it definitely makes no sense that `CanPlay` is true. So in the end MPD's state does not really has to be checked for the `CanPlay` property. It is sufficient to just check whether the play queue is empty.
Please let me know what you think about this pull request if I would amend my commit by applying a patch like this:
```diff
@@ -93,8 +93,8 @@ class Mpris : public QObject
     double MinimumRate() const { return 1.0; }
     double MaximumRate() const { return 1.0; }
     bool CanControl() const { return true; }
-    bool CanPlay() const { return MPDState_Playing!=MPDStatus::self()->state(); }
-    bool CanPause() const { return MPDState_Playing==MPDStatus::self()->state(); }
+    bool CanPlay() const { return MPDStatus::self()->playlistLength()>0; }
+    bool CanPause() const { return MPDState_Stopped!=MPDStatus::self()->state(); }
     bool CanSeek() const { return -1!=MPDStatus::self()->songId() && !currentSong.isCdda() && !currentSong.isStandardStream() && currentSong.time>5; }
     bool CanGoNext() const { return MPDState_Stopped!=MPDStatus::self()->state() && MPDStatus::self()->song()>=0 && MPDStatus::self()->playlistLength()-MPDStatus::self()->song()>1; }
     bool CanGoPrevious() const { return MPDState_Stopped!=MPDStatus::self()->state() && MPDStatus::self()->song()>0; }

@@ -118,7 +118,11 @@ void Mpris::updateStatus()
     if (MPDStatus::self()->state()!=status.state) {
         map.insert("PlaybackStatus", PlaybackStatus());
         map.insert("CanPause", CanPause());
+    }
+    if (MPDStatus::self()->playlistLength()!=status.playlistLength) {
         map.insert("CanPlay", CanPlay());
+    }
+    if (MPDStatus::self()->songId()!=status.songId) {
         map.insert("CanSeek", CanSeek());
     }
     if (MPDStatus::self()->timeElapsed()!=status.timeElapsed) {
```

uhle replied:
Yes, I did. Please see my explanations below.

uhle commented:
Craig, I've rebased the branch to resolve the merge conflict.

uhle commented on dbus/mpris.h:
```diff
@@ -93,11 +93,11 @@ class Mpris : public QObject
     bool CanSeek() const { return -1!=MPDStatus::self()->songId() && !currentSong.isCdda() && !currentSong.isStandardStream() && currentSong.time>5; }
-    bool CanGoNext() const { return MPDState_Stopped!=MPDStatus::self()->state() && MPDStatus::self()->playlistLength()>1; }
-    bool CanGoPrevious() const { return MPDState_Stopped!=MPDStatus::self()->state() && MPDStatus::self()->playlistLength()>1; }
+    bool CanGoNext() const { return MPDState_Stopped!=MPDStatus::self()->state() && MPDStatus::self()->song()>=0 && MPDStatus::self()->playlistLength()-MPDStatus::self()->song()>1; }
```
You are right. Thanks for the hint! And there is yet the same issue with having randomization switched on.\
So the correct solution would be to check whether `MPDStatus::nextSongId()` returns a valid song ID (not equal `-1`).

uhle commented on dbus/mpris.h:
```diff
@@ -93,11 +93,11 @@ class Mpris : public QObject
-    bool CanGoNext() const { return MPDState_Stopped!=MPDStatus::self()->state() && MPDStatus::self()->playlistLength()>1; }
-    bool CanGoPrevious() const { return MPDState_Stopped!=MPDStatus::self()->state() && MPDStatus::self()->playlistLength()>1; }
+    bool CanGoNext() const { return MPDState_Stopped!=MPDStatus::self()->state() && MPDStatus::self()->song()>=0 && MPDStatus::self()->playlistLength()-MPDStatus::self()->song()>1; }
+    bool CanGoPrevious() const { return MPDState_Stopped!=MPDStatus::self()->state() && MPDStatus::self()->song()>0; }
```
Okay, I see. But then I wonder why play queues with exactly one item are treated differently than play queues with two or more items. Does it make sense to have this additional feature available only if there are two items at least?	I've checked what happens if there is only one item in the play queue and 'go previous' is invoked. Then it also starts playing from the beginning of this song.

uhle commented:
Thank you for the review!

> Have you checked these changes work?

Yes, I do test my code changes before I commit and open pull requests. But since I am just a human, it might very well happen that I would not have been aware of all possible test cases in the world. Of course I don't want to introduce regressions. And if there are no regression tests available, the best thing I can do is to judge according to my daily use cases. And obviously, setting 'repeat' or 'random' is something I rarely do. You are so much more experienced with developing Cantata than me because I used to use Cantata since about 2.5 years (and not developing it!) and, yes, I certainly do not use Cantata in its whole complexity. So it's quite natural that you assess the impact of a code change with different eyes.

However, what I have checked so far are the following test cases:
- having one item as well as having more items in the play queue
- having an empty play queue at start and after clearing it (after your first hint)
- initiate playing from `"stopped"` state and monitor what happens when MPD returns to `"stopped"` after finishing to play the last song of the play queue
- switching to different items in the play queue while MPD's state remains to be `"playing"` by clicking on them in random order, incl. clicking on the items at the first and last positions
- playing the song before the last of the play queue and then invoking 'go next' and vice versa by invoking 'go previous', doing the same with the second song and then invoking 'go previous' and vice versa by invoking 'go next' as well as waiting until MPD has finished to play the first song and automatically proceeds with the second one
- moving items of the play queue to different positions, incl. the currently played song to and from the first and last position

After your second hints, I started to play around with setting 'repeat', 'single' and/or 'random' (each alone and combinations of them).

Which test cases am I still missing?

> [...] With which desktop? KDE, GNOME?

Several years ago I switched from GNOME to XFCE because I didn't like some of the decisions taken for GNOME 3 (dynamically adding and removing virtual desktops for instance). XFCE's MPRIS support comes with a panel applet called xfce4-pulseaudio-plugin.

> Are the changes really required? Do they improve the situation, or changed just because?

The current situation without these changes is that the play/pause button of xfce4-pulseaudio-plugin is always disabled because of
https://github.com/CDrummond/cantata/blob/5e5b362130c93eaa4d8736f838ec659782496c72/dbus/mpris.cpp#L120-L121
That means that I also cannot use the media keys to pause and resume playing unless I switch the focus to Cantata's main window and use its controls. To me this is simply annoying. IMHO I would regard this as being a bug. Once I have found the culprit, I realized after removing the comment signs that `CanPlay`/`CanPause` is always enabled and thought that this doesn't really make sense. Any other thoughts are briefly described here: https://github.com/CDrummond/cantata/pull/1592#discussion_r413272338 \
I also wondered why `CanGoNext` and `CanGoPrevious` are still enabled if the currently played song is on the last or the first position of the play queue respectively, since I don't know such behaviour from other media players. Now that I've understood that invoking 'go previous' also has this additional feature to go to the beginning of the current song, I see indeed that `CanGoPrevious` should be true unless MPD's state is `"stopped"`. But why then checking whether there are two or more items in the play queue? What is the rationale? The case with `CanGoNext` is different though, but I have seen that simply checking `MPDStatus::nextSongId()` for a valid song ID yields the desired result in every situation that I have tested so far (incl. any repeat or random settings). Anyway, I decided to revert those changes because I have seen that the same conditions are applied for the controls in the tray of Cantata's main window and I don't want to introduce inconsistent behaviour.

If you would agree, I would like to propose to adapt the conditions to enable/disable the 'go previous'/ 'go next' controls in a separate pull request.

While playing around with 'repeat' and 'single' settings, I realized that `LoopState` is not respecting `MPDStatus::single()`. `LoopState` shall equal `"Track"` according to the [MPRIS interface specification](https://specifications.freedesktop.org/mpris-spec/latest/Player_Interface.html) if both, `MPDStatus::repeat()` and `MPDStatus::single()`, return true. I changed `Mpris::LoopState()` and adapted `Mpris::SetLoopState()` accordingly.

_Do all these changes improve the current situation?_ Yes, I think so. At least I hope that they don't make the situation worse.

> I only ask as it was a pain getting to work correctly before, and no-one has filed a bug report to state that it doesn't work as currently implemented.

Understood. If it's that what you want, I still have a few issues on my list...  But since you have stated more than once that you only have limited time to spend into bug fixing of Cantata and will no longer develop it w.r.t. to new features (what I completely understand) and I am quite skilled in developing Qt applications on the other hand, so I just thought, let's be kind and help you out by trying to find the reasons for some issues myself and to propose appropriate fixes. I wouldn't do this if I would not like Cantata that much.

> Also - only 1 changelog entry is required. e.g. "Fix various MPRIS issues" - or word to that effect.

Have done so.

CDrummond commented:
Thanks for all of this, I do appreciate it. Cantata's codebase is not the best in the world...
